### DISCUSSION TITLE
Final ESP32 design completely replaces Tuya TYWE1S Chip

### DISCUSSION THREAD

**Discussion by jazzmonger** [jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip#34] [Announcements]
What a loooong road.  But perseverance, tenacity and a bit of luck paid off.  Here's the final working version mods fully documented.  If you're not comfortable modifying circuit boards, I don't recommend trying this.  If you are, you're in luck!

I'll start with the main MCU Board.  It has 2 mods.  We need to break the connection between the MCU Auger motor signal and the opto-isolator that controls the Auger motor.  I use a razor blade to cut the trace and scrape the insulation away from the traces so you can solder wires to the break.  You should also run a ground wire, just to be safe.  I picked one up where the black wire is.

![image](https://user-images.githubusercontent.com/52110065/225821001-876dfed2-2c02-4def-830d-1c2a38b2c21d.png)

Then these 2 wires are run up to the display.  I use a USB cable and plugs you can get cheap on amazon.  It's shielded and works great.  You could hard wire the cable, but I like to be able to unplug stuff.  You never know when you might need to unplug something. Hot. glue works great for affixing all this extra stuff.

<img width="857" alt="image" src="https://user-images.githubusercontent.com/52110065/225821151-fa122fd8-b35e-4012-bfd9-8f9453aef549.png">

Next I modify the Display board.  We need to tap both the TX and RX RS232 signals going to and from the MCU and Display so we can see whats happening between the 2 modules. I remove and then replace the plastic molex 4 pin plug and tap in from the top as the back pins are blocked by the LCD matrix.

![image](https://user-images.githubusercontent.com/52110065/225821622-4b5e0e86-2b2a-4337-8563-9e5203bddc84.png)

Most of the WiFi module pins aren't being used for anything, so I repurpose them for these mods.

![image](https://user-images.githubusercontent.com/52110065/225822001-53c41d7d-2430-4be5-bae1-fd8d2d825125.png)

I also repurpose the useless Timer LED to indicate Ultra Low Power mode.  All MCU signals are 5v, so we need to use a lower value 470ohm resistor so the LED is brighter.

If there was ever a time to fabricate a custom circuit board, this is one of them...  But we have to prove the design works first!  So, here goes...
<img width="1433" alt="image" src="https://user-images.githubusercontent.com/52110065/225822701-8686c593-dfc3-40de-aee1-3ab0ae58d87c.png">

D1 Mini ESP32, Bidirectional 4 channel Level shifter (3.3v to 5v), perf board and LOTS of patience soldering little wires. Heres the final wiring designations:

-  GPIO16 <- TX PIN ON DISPLAY BOARD
-  GPIO21 -> RX PIN " " "
-  GPIO27 <- MCU CONTROL BOARD SENSOR FROM MCU
-  GPIO25 -> AUGER OUT TO OPTOISOLATOR
-  GPIO19 -> LED ON DISPLAY FRO P5
-  GPIO26 -> THUMBS UP RESET  
-  GPIO5   <-> SDA
-  GPIO33 <-> SCL

I use 4 pin plugs for the Time of Flight pellet level sensor and the USB jack.  This way i can completely remove the wifi module if I need to make mods.  All could be hard wired
![image](https://user-images.githubusercontent.com/52110065/225823203-f9532fb3-f9fc-47ec-adf6-58c34d0c79b3.png)

<img width="1528" alt="image" src="https://user-images.githubusercontent.com/52110065/225823276-33cb83e1-9f5c-474d-bcb7-9d5448c668cf.png">
 
And everything fits nicely under the cover
<img width="1132" alt="image" src="https://user-images.githubusercontent.com/52110065/225823448-df9a5551-411c-45c8-8722-7885b34713db.png">

A Dremel helps cut holes and trim the end of the display board so the USB jack and wires fit ok.

I built all of this in one sitting, about 4 hours including testing.  That does NOT include the entire YEAR I've been working on all of this!

Next, you flash the ESP32 with my ESP32 firmware.  Adopt it in ESPHome and have a party!

The Pellet Stove Command Center.  Everything you wanted in a Pellet Stove but were afraid to ask.

![image](https://user-images.githubusercontent.com/52110065/225824290-c2578637-e9a0-420f-a67c-9d3871ac1b42.png)

The values I show for all the sliders seem to keep the pot lit at a very low burn rate.  If it tends to go out, play with the values.  Its very therapeutic sitting next to a pellet stove with your iPad or laptop and studying how burning pellets act.  Everything can be adjusted depending on what you want.  Future versions might self adjust depending on the outside temperature and how hard the stove needs to work to maintain the temp of the room/house.  Normally the pot burns at 275-300C, even on the lowest P4 setting.  Lowering it to 180-190C makes a HUGE difference.  
<img width="1182" alt="image" src="https://user-images.githubusercontent.com/52110065/225826380-a7518656-6924-468a-a44d-e3bcbdbe8d71.png">
I've had this thing burning solid for almost a week at times, only turning it off briefly to clean it. Also, because it's not constantly cycling on/off, pellet burn is WAY down and the glass stays mostly clear.

I'v also set the cleaning cycle on the display to 15s at 1 min intervals (settings/thumbs up until you get to this setting).  This helps keep the embers going and light new pellets a little faster as they drop.

**Comment by bpohvoodoo** [comment]
I'm very interested in rebuilding this project. Do you have the schematics of the wiring? I do not want to mess up the wiring.

**Reply by jazzmonger** [reply]
Well...winter is almost here so I'll be revisiting this project in a couple of weeks.  I actually want to create a circuit board for it.  I have to get better with KiCad tho... I've only done discrete designs so far, nothing using ESPs.

I see you're in Germany... what model of stove do you have? You might want to start with the basics first and reflash your Tuya board with my basic ESPHome firmware.

**Comment by bpohvoodoo** [comment]
Yes i am in Germany. I have a Nemaxx P12 here. Is it possible to backup the original firmware? Maybe i can replace the Tuya  to preserve the firmware.

As an Alternative to cutting the wire and add the 470Ohm resistor for the timer led, can i replace the smd 1k resistor with a 470 Ohm smd resistor?

**Comment by jazzmonger** [comment]
It is possible, Google is your friend here!The first step is to get Comms working with the chip.  Then just about anything is possible.

>Yes i am in Germany. I have a Nemaxx P12 here. Is it possible to backup the original firmware? Maybe i can replace the Tuya  to >preserve the firmware.

**Comment by bpohvoodoo** [comment]
Want to try the Basic version, but i've got two problemes:
compiling esphome ends with:
INFO ESPHome 2023.9.3
INFO Reading configuration /config/esphome/stove.yaml...
INFO Updating https://github.com/esphome/esphome.git@pull/4032/head
INFO Detected timezone 'Europe/Berlin'
INFO Generating C++ source...
INFO Compiling app...
Processing wood-pellet-stove (board: esp01_1m; framework: arduino; platform: platformio/espressif8266@3.2.0)
--------------------------------------------------------------------------------
HARDWARE: ESP8266 80MHz, 80KB RAM, 1MB Flash
Dependency Graph
|-- ESPAsyncTCP-esphome @ 2.0.0
|-- ESPAsyncWebServer-esphome @ 3.1.0
|-- DNSServer @ 1.1.1
|-- ESP8266WiFi @ 1.0
|-- ESP8266mDNS @ 1.2
|-- noise-c @ 0.1.4
Compiling .pioenvs/wood-pellet-stove/src/esphome/components/tuya/tuya.cpp.o
Compiling .pioenvs/wood-pellet-stove/src/esphome/components/uart/uart_component_esp_idf.cpp.o
Compiling .pioenvs/wood-pellet-stove/src/esphome/components/uart/uart_component_libretiny.cpp.o
Compiling .pioenvs/wood-pellet-stove/src/esphome/components/uart/uart_component_rp2040.cpp.o
Compiling .pioenvs/wood-pellet-stove/src/esphome/components/uart/uart_debugger.cpp.o
Compiling .pioenvs/wood-pellet-stove/src/esphome/components/uptime/uptime_sensor.cpp.o
src/esphome/components/tuya/tuya.cpp: In member function 'void esphome::tuya::Tuya::send_local_time_()':
src/esphome/components/tuya/tuya.cpp:465:9: error: 'ESPTime' is not a member of 'esphome::time'; did you mean 'esphome::ESPTime'?
  465 |   time::ESPTime now = time_id->now();
      |         ^~~~~~~
In file included from src/esphome/components/time/real_time_clock.h:8,
                 from src/esphome/components/tuya/tuya.h:9,
                 from src/esphome/components/tuya/tuya.cpp:1:
src/esphome/core/time.h:12:8: note: 'esphome::ESPTime' declared here
   12 | struct ESPTime {
      |        ^~~~~~~
src/esphome/components/tuya/tuya.cpp:466:7: error: 'now' was not declared in this scope; did you mean 'pow'?
  466 |   if (now.is_valid()) {
      |       ^~~
      |       pow
*** [.pioenvs/wood-pellet-stove/src/esphome/components/tuya/tuya.cpp.o] Error 1
========================== [FAILED] Took 1.87 seconds ==========================

And how can i install/add the basic card?

**Comment by jazzmonger** [comment]
That’s not an error that I am familiar with. It looks like some kind of problem with your time zone and/or country. You might try posting a question on the ESPHome Discord and ask about the issue or on the ESPhome forums. 

bpohvoodoo ***@***.***> wrote:﻿
Want to try the Basic version, but i've got two problemes:
compiling esphome ends with:
INFO ESPHome 2023.9.3
INFO Reading configuration /config/esphome/stove.yaml...
INFO Updating ***@***.***/4032/head
INFO Detected timezone 'Europe/Berlin'
INFO Generating C++ source...
INFO Compiling app...
Processing wood-pellet-stove (board: esp01_1m; framework: arduino; platform: ***@***.***)
HARDWARE: ESP8266 80MHz, 80KB RAM, 1MB Flash
Dependency Graph
|-- ESPAsyncTCP-esphome @ 2.0.0
|-- ESPAsyncWebServer-esphome @ 3.1.0
|-- DNSServer @ 1.1.1
|-- ESP8266WiFi @ 1.0
|-- ESP8266mDNS @ 1.2
|-- noise-c @ 0.1.4
Compiling .pioenvs/wood-pellet-stove/src/esphome/components/tuya/tuya.cpp.o
Compiling .pioenvs/wood-pellet-stove/src/esphome/components/uart/uart_component_esp_idf.cpp.o
Compiling .pioenvs/wood-pellet-stove/src/esphome/components/uart/uart_component_libretiny.cpp.o
Compiling .pioenvs/wood-pellet-stove/src/esphome/components/uart/uart_component_rp2040.cpp.o
Compiling .pioenvs/wood-pellet-stove/src/esphome/components/uart/uart_debugger.cpp.o
Compiling .pioenvs/wood-pellet-stove/src/esphome/components/uptime/uptime_sensor.cpp.o
src/esphome/components/tuya/tuya.cpp: In member function 'void esphome::tuya::Tuya::send_local_time_()':
src/esphome/components/tuya/tuya.cpp:465:9: error: 'ESPTime' is not a member of 'esphome::time'; did you mean 'esphome::ESPTime'?
465 |   time::ESPTime now = time_id->now();
|         ^~~~~~~
In file included from src/esphome/components/time/real_time_clock.h:8,
from src/esphome/components/tuya/tuya.h:9,
from src/esphome/components/tuya/tuya.cpp:1:
src/esphome/core/time.h:12:8: note: 'esphome::ESPTime' declared here
12 | struct ESPTime {
|        ^~~~~~~
src/esphome/components/tuya/tuya.cpp:466:7: error: 'now' was not declared in this scope; did you mean 'pow'?
466 |   if (now.is_valid()) {
|       ^~~
|       pow
*** [.pioenvs/wood-pellet-stove/src/esphome/components/tuya/tuya.cpp.o] Error 1
========================== [FAILED] Took 1.87 seconds ==========================
And how can i install/add the basic card?

**Comment by bpohvoodoo** [comment]
I doent think it is a timezone problem. Looks like function definitions in librarys changed.
Can you please try to compile your basic version?

**Comment by jazzmonger** [comment]
As soon as it gets cold here in a week or so I’ll be revisiting all of it, so, yes!

>I doent think it is a timezone problem. Looks like function definitions in librarys changed.
>Can you please try to compile your basic version?

**Comment by jazzmonger** [comment]
Ok, I figured it out. Remove these lines and it will compile.  I just got the same error on mine and now it works just fine.
```
external_components:
 - source: github://pr#4032
    components: tuya
    refresh: 0s
```
The devs had created a Tuya fix (hack) for me quite a while ago.  It got merged with the ESPHome core, so declaring it again steps all over it.

**Comment by bpohvoodoo** [comment]
Nice, but im building your Project with ESP32 now. In looked in your code and wired the gpios. Used an Sparkfun ESP32 Thing. A wirering diagram would be nice, but i’m confident to get it working without one ;-). I’ve got only one problem left. How do I add the lovelace card. Tried different methods, but none is working.

**Reply by jazzmonger** [reply]
Did you get the Tuya chip to flash?  I'd at least start there so you get a feel of how all this stuff works.
Lovelace can be tricky for sure. Make sure you install Plotly Graph Card in HACS.  I can probably break down each object into code bits if needed.
Are you using a logic level changer on the gpio inputs/outputs?
Can you get the ESPHome object working? If it compiles and uploads,it should be discovered and you can add it to Lovelace,
I'm happy to help in any way I can. I'd say mine will be in full swing use in a couple of weeks once it gets cold here. 

I had a look at the data sheet and it looks like that chip should work ok. 
https://www.espressif.com/sites/default/files/documentation/esp32_datasheet_en.pdf
The biggest issue I had was many of the esp32 variants (S2, S3, C2,C3,etc) don't support 3 full UARTs.  Yours does. I also chose the "esp32 mini" bc of its size. It's tiny and fits inside the control module perfectly.  You might have some challenges there.

**Comment by bpohvoodoo** [comment]
No. I didn't flash the tuya chip, because I want to preserve the original module.
Yes, I'm using 2 x 4-channel level shifter for RX-TX, etc. Because the OnBoard MCU uses 5V levels. Used it for the timer-led too, so i don't have to change the resistor. Want to change as less as possible on the original board.

Which ToF-Sensor do you use?
GY-VL53L0XV2 I2C IIC ?

**Reply by bpohvoodoo** [reply]
![image](https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/assets/4918785/998e87e7-9f00-49ec-b299-1bad4bd35e53)

**Reply by jazzmonger** [reply]
looks good!  And yes, that's the TOF sensor- VL53L0X.  it works really well, rock solid and highly accurate.
I fired my stove up yesterday.  All is working well! 

I just started on the circuit board design as well.  I found a guy who built a basic esp32 Dev board on PCBWAY and is highly proficient in KiCad and shared his board designs and schematics, so that gave me the kick in the butt I needed to do a circuit board! Watching his video is nothing short of mesmerizing!

https://www.pcbway.com/project/shareproject/iot_thing_universal_iot_development_board_fe54c4ad.html

It's not as scary as I thought it would be.... I've never done a 4 layer board before and he makes it all make sense, it's all in defining all the nets and the board setup.

 I also realized that bidirectional level shifters work great but are overkill.  Simple logic Buffers will suffice.  All the 3<->5v signals are unidirectional.

I'll start another discussion on that whole process.

**Comment by bpohvoodoo** [comment]
Hope this is finally correct as i tried to guess the socket pinout from your fotos.

![image](https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/assets/4918785/05646a2a-da0e-491a-bb7e-250bc2029738)

**Reply by jazzmonger** [reply]
Looks good.  Keep in mind that I ran jumpers from the display rx/tx lines, led, etc. and routed them to unused pins on the daughter board socket.

![IMG_2559](https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/assets/52110065/c4e3c7b5-efa5-42f7-8873-9f3496e95098)
 

![IMG_2563](https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/assets/52110065/a5f42997-022d-4326-b5c0-28cbbc95e505)
![IMG_2562](https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/assets/52110065/b86e631f-2e35-4a7c-a9b1-6c4b9cf58e5c)
to J5.

**Comment by bpohvoodoo** [comment]
Ok, i noticed the Yellow and Red wires, but where do those two blue wires go?

**Reply by jazzmonger** [reply]
The top one taps the LED.  I cut the trace to it from the MCU so I could repurpose it. The bottom one was my first pass at tapping the display tx on the back (solder) side of the 4 pin molex connector under the lcd.  That was a pain so I redid it using the red and yellow wires shown in the pic above by pulling off the plastic from the connector, soldering the wires and putting the plastic part back.

Theres also a reset wire that taps the side of the thumbs up button that isn't ground. When the stove throws an error (these stoves throw lots of errors), the only way to clear it is to hold the thumbs up button for 2 sec.  I emulate that button press using a GPIO signal thru a level shifter.
Then expose it as a button in my UI so I can clear errors without touching the stove.
![IMG_2564](https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/assets/52110065/526e6300-9f73-4542-8952-6dd503455a1d)

clearly I have way too much time on my hands!

**Comment by bpohvoodoo** [comment]
Thank you for your explanation.
But there are some questions left. I realised that you have used only a 4 channel level shifter.
You seem not to use level shifter for the Opto-Isolator, for one RX line and for led. 
For the led it is clear, because you used an alternative resistor,
but what are the correct levels for the opto isolator? What is the input level. It seems to be directly connected to the MCU output (5V), but in the specs of the isolator it says 3V input level. I’m confused about that.
It is possible that you make a littel drawing of the J5 connector (16-pin) with your pinout?

**Reply by jazzmonger** [reply]
When I measured the optoisolator, the signal levels were already 3v and this is confirmed by the spec sheet.
![IMG_2565](https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/assets/52110065/08a0f99e-0643-4820-a323-5efcb9be4ec7)

so neither of those 2 signals need to be level converted as they are already 3v. And you understand the LED part. Frankly I only use the level shifter bc I feel it's good design practice.  Plus the reset output function MCU pin REQUIRES a 5v input signal, but all the other pins are monitoring signals only. The ESP32 chips can handle 5v input signals all day long.  This has been widely debated but put to rest specifically by a statement made by the CEO of Espressif who says go ahead and use 5v input signals with his ESP chips.
![IMG_2566](https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/assets/52110065/66b5bb67-c1f4-4904-b949-be4525abe8a9)

as far as a drawing, I'm creating a 4 layer pc board for all of this in KiCad, so I will be doing a schematic.  I'm busy right now winterizing my house and RV - winter is almost here.  Once it gets cold, I'll be inside spending lots of time in KiCad!

**Comment by bpohvoodoo** [comment]
my news: Builded prototype and changed some datapoint for Nemaxx p12 and now it is working. ;-)
Are the exhaust temps in your sketch in F or °C. I think in F, or am I wrong. Because mine is sendig °C.

**Reply by jazzmonger** [reply]
That's very Cool!  Did you get the Lovelace cards going?  So, temps for "most" objects in ESPHome originate in C.  If you're in th EU and specify C in your home assistant config, then everything should be in C.  I'm in the US and we use F in HA, so that's what all my temps  show in my graphs.

This is complicated as my stove reports data points in F. So in my case HA expected C, but gets F and converts F to F.  Not good... So the devs made a modification to the climate entity with a new entry for me as follows:

![IMG_2567](https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/assets/52110065/cecb780a-8918-4ce1-a953-46487bc1edfe)

Since my stove reports all TUYA temp datapoints in F, I set this to TRUE.  You should set it to FALSE if your stove is already reporting in C. Make sense?

Also, it would be good to know what changes you had to make in the dp's for your stove so the next victim has an easier time!

I am having my morning coffee sitting next to my pellet stove watching my "Pellet Stove Command Center".  I can sit here for hours watching all the sensors and graphs, tweaking a control value here or there, and stay nice and warm but not get burned out of the house or inundated with errors and failures!
 It still amazes me just how cool ESPHome and home assistant are when working together like they do in this project. I'm not exaggerating whenI say I've got hundreds of hours invested in this project. I'm also glad someone like you randomly decides to duplicate it and it works.  Clearly a circuit board will make it easier....  I have another pellet stove in my shop waiting for that!
Jeff

**Reply by bpohvoodoo** [reply]
I had set set the above variable to false already, but wasn't shure if all the temps in the automations and yaml files were in F. Had to convert it to make stoking work, because mine never came to such exhaust temps. 
Here is my Card. As i haven't made the modifications on the display board. Some things do not work at the moment.
At the moment of screenshot the ESP was offline as i am working on it ;-)
![image](https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/assets/4918785/8960ae68-6a3b-41a3-b8d0-be26827c009d)
You can see all my changes in my fork.

**Reply by jazzmonger** [reply]
You are very close! This will be great once you're done. I didn't anticipate someone using only Celsius!  

Here are my current settings - these will get your pellet burn close once your fire it up.
![IMG_2570](https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/assets/52110065/bcc28aeb-0201-4eec-b4bc-462c0f358ecc)

Btw, sometimes it looks like HA misses data points reported by the esp,  and it results in anomalies like this
![IMG_2572](https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/assets/52110065/7e64fdc5-8637-4767-8a6c-9bff72cec4a0)

Those pulses are actually there, just not showing on the graph, even though he data is in the ha db.  I finally tracked it to the reporting add-on.  The Dev of that Plotly plug-in is also stumped as to why this happens. It's a nit, and most of the time the graphing of that add-on rocks, so I just live with it. Here's that card:

```
type: custom:plotly-graph
defaults:
  entity:
    show_value: true
hours_to_show: 0.3
entities:
  - entity: sensor.house_stoking_sensor
    show_value: 0
    filters:
      - filter: y !== null && +y > 100
    yaxis: y1
    line:
      color: red
      dash: dashdot
  - entity: sensor.house_auger_display_sensor
    show_value: 0
    filters:
      - filter: y !== null && +y > 90
    yaxis: y1
    line:
      color: aqua
  - entity: sensor.house_uart_pipe_temp
    name: UART pipe temp
    filters:
      - median:
          window_size: 3
          extended: 0
          centered: true
    yaxis: y1
    line:
      color: yellow
      shape: spline
      width: 2
      smoothing: 0
   ```

**Comment by bpohvoodoo** [comment]
Almost getting it. Made all mods. P5 Led - check, Thumbs Up Reset -> check
But with the mainboard Auger Motor mod there is a problem left. The D1 Mini32 is not getting GPIO27 to high. So the motor is running permanently. I had to bridge the GPIO25 and 27. To make it work temporarily. Maybe my module is defective. I have to figure this out.
By the way:  is  sensor.house_stoking_sensor equal to stoking_flag? I do not have a stoking_sensor entity. Entity sensor.house_auger_display_sensor is missing, too. Trying to debug the display_text_sensor messages, because in my case what you see is not what you get ;-).
On picture you see "Auf Widersehen!" on display there is "ECO1" Added the hexcode to the message.
![image](https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/assets/4918785/4d45b4f3-e64f-4df0-bd42-9517dcaa222a)
Waiting for my ToF Sensor.

**Reply by jazzmonger** [reply]
You are taxing my little pea brain... I had to go thru all my code to answer this question!  Funny how we solve a problem in code and then move on and forget what we did at the time. 

binary_sensor.house_stoke_fire_flag is just a binary sensor.  When stoking the fire, it gets set to TRUE in ESPHome:
```
  - platform: template # template sensors expose sensors to HA
    name: ${location} Stoke Fire Flag #<<--HA exposed name
    id: stoking_flag #<< internal ESPHome identifier
    # set it to the value of the global
    lambda: |-
      if (id(stoking_flag_global) == 1) {
        return true;
      } else {
        return false;
      }
```
Then,  I created a HA template sensor to help display the state of this sensor in a graph - its purely visual. The upper portion of the graph is the temp that the stoking started (as set by a helper slider), the lower portion is 130 which I chose arbitrarily so the graph looks good.  It could just as easily been zero, but the auger pulses take up the 100-120 range in the graph and they are defined in a similar way.  Graphing 1/0 or ON/OFF flags proved challenging with Plotly so this is the "hack" I used to get around that limitation.  Plotly has so many other nice attributes, I forgive it for this shortcoming.

```
  - name: 'House Stoking Sensor'
    state: >
        {% if states('binary_sensor.house_stoke_fire_flag') == 'on' %}
          {{ states('input_number.house_stoking_temp') }}
        {% else %}
          130
        {% endif %}
```
![image](https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/assets/52110065/17b4b12b-d27e-4ba1-a086-ae2cdbac6625)
you can see that when the exhaust temp hits 176F (set by a helper slider), the fire is likely going out.  So, stoking routines start.  The Auger ON/OFF time during stoking changes to get more pellets into the pot so the fire doesn't go out. When exhaust temp goes back above the stoking temp you set, stoking stops and the flag is set to false.  These routines really are the magic that lets the burn pot just cruise along at a very low temp when in  IDLE mode. When in HEATING mode, the auger signals are just echoing P1/P2/P3/P4, whatever you have it set for.

_The D1 Mini32 is not getting GPIO27 to high_
interesting.  Not all GPIO outputs are created equally... I've had to finesse several of them in the past when I saw them misbehave.  What about just using GPIO25?  

_Trying to debug the display_text_sensor messages, because in my case what you see is not what you get ;-)._

Dude, if I told you just how many hours I spent trying to decode and get all those messages to be correct.... They are "mostly" right.  This stove does some REALLY weird and obtuse shit.  Good for you if you can get them to be 100% accurate!  At least I did all the heavy lifting by figuring out the HEX decoder!

From what I can tell, you are 99% of the way to the finish line!

**Comment by bpohvoodoo** [comment]
So, my auger control mod is now working, too. I figured out, that the potential is 5V between Pin 2 (Cathode) of MOC3021 and GND. So i had to use a level shifter. Now it works.
Last Problem P5 does not kick in.

**Reply by jazzmonger** [reply]
Yes, all MCU inputs as well as outputs are all  5V.  ESPs can handle 5v inputs, but cannot support 5v outputs w/out level shifters!

P5 is"Ultra Low".  it only kicks in when the stove goes from "Heating" to "Idle".  You can force it into Idle by just turning the set temp of the stove to below the current room temp.  That forces the stove into "idle" mode once its lit and running.  Also, for this to work, you MUST be in ECO2 mode on the stove.

ECO1 mode on the stove turns it ON when room temp < SET Temp, lights the pellets and then heats the room to SET temp.  Then it shuts down the fire, cools, and the cycle starts over.

ECO2 is much different.  Unmodified (before our mods), the stove turns on, lights and then runs at P1/P2/P3/P4 or whatever you have it set for while the stove is in Heating mode.  Once the room temp = SET temp, the stove shifts from Heating mode into Idle mode and shifts down to P4 until the room temp goes below SET temp.  Then it shifts to P1/P2/P3/P4 or whatever its set for.  I generally set the stove to P2 or P3 at all times since when using P1 the over temp sensor is often tripped and throws an error. ECO2 mode sounds good in theory, but the room often overheats and doesn't stop heating until you turn off the stove after sweating bullets.

While in ECO2 mode and _**when**_ the stove shifts to Idle mode, if the "Ultra Low" or P5 toggle is on, then that's when all my routines take over.  You can watch the "Auger sensor from MCU" signal toggle on/off/on/off/etc.  It should be coupled to the "Auger out" signal while lighting and Heating mode are active and follow it exactly.  Then when shifting to Idle mode w/ P5 on "Auger Out" uncouples from "Auger Sensor from MCU" by executing my routines and turns on/off/on/off/etc according to the slider settings for "Auger On Time" and "Auger Off Time".   If the fire almost goes out, stoking occurs, but only in ECO2 mode and only while the stove is Idling.

Hopefully this clears things up.

Jeff

**Reply by bpohvoodoo** [reply]
Found the problem. Thought to turn on Indicator switch will trigger P5, put had to create a helper.

**Reply by jazzmonger** [reply]
How's it working???

**Reply by bpohvoodoo** [reply]
Hi,

it is working ok, but the ToF Sensor crashes every few days. Have to reset ESP for it starting to work again.
Still working on Debuging UART, but no new things found.
It seems like the display itself manages the most of the functions.
Sometimes the stove is beeping, which is an indication, that some datapoint values are send, but i have no automation running and nothing entered in HA.

Kind regards,
Mirco

**Reply by jazzmonger** [reply]
Which TOF sensor are you using?
I use this one VL53L0X, and I have literally never had a problem with it. I had all kinds of issues with the ultrasonic sensors, where it would randomly stop sending updates every couple of days and it required to restart of the esp. 

UART Debugging???  
Are you trying to get additional messages to display?

**Reply by bpohvoodoo** [reply]
ToF:
If use one of those: https://www.amazon.de/dp/B086TSKJBT?psc=1&ref=ppx_yo2ov_dt_b_product_details as ToF sensors.
But I would prefer Adafruit VL53L4CX Time of Flight Distance Sensor - ~1 to 6000mm - STEMMA QT / Qwiic, but not possible with ESPHome at the moment.

UART:
I still have the problem, that the display is not in sync with the text sensor of ESPhome. This is mostly occuring if you are rebooting the ESP (ToF Problem) and on stove power on. My idea is to pass thru the diplay MCU serial command (like auger sensor) and to find a command which puts the stove into a defined state.
My questions are:
What is the difference between total POWER OFF and ECO1, because my text sensor never gives me ECO1.
Can I send settings to MCU (P1,P2,P3,P4 auger on/off times) and others.
to be continued.
For this to have a chance to work I will need a TX pin for UART3.

**Reply by jazzmonger** [reply]
we're using the same [Sensor](https://www.amazon.com/gp/product/B09VB5KDV2/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1)

I gave up trying to get the display in sync after ESP restarts via the TX UART messages.   Its next to impossible... I know when the stove is off it's "Goodbye" and can test for when it's "Idle" or "Heating" or "Off".  So, I manually force those messages after a restart. You can see that in the ESPHome code.  That's about the best you can hope for.  The stove doesn't know or even care when the ESP is there or not, so there's no update messages to look for to set the display after a restart. 

And ECO1 is a mode that is mostly useless. In ECO1 mode, 50%++ of the time the stove fails to light the pellets and throws an error. In fact, this is exactly what led me to create this entire project! Once the stove lights in ECO2 mode and shifts into "ultra low power" mode, it can idle along for literally days as long as pellets are refilled. And Alexa tells me when to fill the pellets thru an automation that checks the TOF sensor. I've had it on non-stop for over 2 weeks at one point when it was really cold here. I use Automations to light it up when the room temp is < 63F and turn it off > 70F. Always in ECO2 mode.

you use the select:select service to chose the power level and ECO1/ECO2 modes:

in an Automation:
```
service: select.select_option
data:
  option: P3-Med Low
target:
  entity_id: select.house_power_select
```
In ESPHome, they are defined:
```
select:
  - platform: "tuya"
    tuya_id: "${location}"
    name: "${location} Power Select"
    id: power_select
    enum_datapoint: 4
    options:
      0: P1-High
      1: P2-Med High
      2: P3-Med Low
      3: P4-Low      
    on_value: #store last value so it can be restored on power/boot up or ESP reset
      - lambda: id(power_select_lastvalue) = i;
      
  - platform: "tuya"
    tuya_id: "${location}"
    name: "${location} ECO Select"
    id: eco_mode
    enum_datapoint: 101
    options:
      0: ECO1 - Heat then OFF
      1: ECO2 - Px then P4
```

**Comment by mannyd209** [comment]
Did you ever make your own PCBs? I'm looking to buy one of these pellet stoves soon and want full control in HA without getting too crazy on it. LoL

**Reply by jazzmonger** [reply]
I’m thinking of doing one soon. The ESP in my shop stove keeps disconnecting from Wi-Fi, although it runs the stove perfectly fine, the connection only stays valid for a few minutes and then I have to reset it. So the WiFi chip on the D1 is failing since all other 50 devices in my shop have no issues. 

I don’t think it’s going to be a big deal since I’m going to use an ESP 32 D1 mini to start with instead of designing it from scratch.  That way, if the ESP ever fails in the future, you can just unplug the D1 and plug a new one in. 

So it’s really just a couple of headers, a level shifter, jack holes,  mounting holes and circuit traces. 

Edit:  I've started :)
![image](https://github.com/user-attachments/assets/8a9b34b1-2f56-440c-862d-4000fcbfdcd9)


I'm no expert in KiCad, but I can stumble my way thru.  I'm only going to do a PCB that accommodates off the shelf modules since I'm not keen on debugging a full blown SMT ESP32 board.  Mount a WeMOS ESP32 D1, a level shifter and a few JST connectors (or just wires) and you're off to the races. 

![image](https://github.com/user-attachments/assets/00948392-4b57-48ca-9dcc-0e939eb69d78)

I'll update my progress in a new discussion.

**Reply by jazzmonger** [reply]
More progress on the circuit board
https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/discussions/41

**Comment by jazzmonger** [comment]
PID controller is now working!  Temps go even lower = better control.
[PID](https://github.com/jazzmonger/wood-pellet-stove-with-TYWE1S-Tuya-chip/discussions/42)
